sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessorBase.cs

```diff
@@ -0,0 +1,729 @@
+            // partitions it owns.  When an event processor goes down and it has only expired ownership, it will not be taken into consideration
+            // by others.
+
+            var partitionDistribution = new Dictionary<string, int>
+            {
+                { Identifier, 0 }
+            };
+
+            foreach (PartitionOwnership ownership in activeOwnership)
+            {
+                if (partitionDistribution.TryGetValue(ownership.OwnerIdentifier, out var value))
+                {
+                    partitionDistribution[ownership.OwnerIdentifier] = value + 1;
+                }
+                else
+                {
+                    partitionDistribution[ownership.OwnerIdentifier] = 1;
+                }
+            }
+
+            // The minimum owned partitions count is the minimum amount of partitions every event processor needs to own when the distribution
+            // is balanced.  If n = minimumOwnedPartitionsCount, a balanced distribution will only have processors that own n or n + 1 partitions
+            // each.  We can guarantee the partition distribution has at least one key, which corresponds to this event processor instance, even
+            // if it owns no partitions.
+
+            var minimumOwnedPartitionsCount = partitionIds.Length / partitionDistribution.Keys.Count;
+            var ownedPartitionsCount = partitionDistribution[Identifier];
+
+            // There are two possible situations in which we may need to claim a partition ownership.
+            //
+            // The first one is when we are below the minimum amount of owned partitions.  There's nothing more to check, as we need to claim more
+            // partitions to enforce balancing.
+            //
+            // The second case is a bit tricky.  Sometimes the claim must be performed by an event processor that already has reached the minimum
+            // amount of ownership.  This may happen, for instance, when we have 13 partitions and 3 processors, each of them owning 4 partitions.
+            // The minimum amount of partitions per processor is, in fact, 4, but in this example we still have 1 orphan partition to claim.  To
+            // avoid overlooking this kind of situation, we may want to claim an ownership when we have exactly the minimum amount of ownership,
+            // but we are making sure there are no better candidates among the other event processors.
+
+            if (ownedPartitionsCount < minimumOwnedPartitionsCount
+                || ownedPartitionsCount == minimumOwnedPartitionsCount
+                && !partitionDistribution.Values.Any(partitions => partitions < minimumOwnedPartitionsCount))
+            {
+                // Look for unclaimed partitions.  If any, randomly pick one of them to claim.
+
+                IEnumerable<string> unclaimedPartitions = partitionIds
```

Comment:
I'd consider making this `ToArray` call, since you've got a limited size and are using the positional properties of it multiple times below.